File: ios/OctopusEventSerializer.swift
import Foundation
import Octopus

/**
 * Serializes OctopusEvent objects to dictionaries for React Native bridge.
 */
class OctopusEventSerializer {
  
  static func serializeEvent(_ event: OctopusEvent) -> [String: Any]? {
    switch event {
    case .postCreated(let context):
      var contentList: [String] = []
      if context.content.contains(.text) { contentList.append("text") }
      if context.content.contains(.image) { contentList.append("image") }
      if context.content.contains(.poll) { contentList.append("poll") }
      return [
        "type": "postCreated",
        "postId": context.postId,
        "content": contentList,
        "topicId": context.topicId,
        "textLength": context.textLength
      ]
      
    case .commentCreated(let context):
      return [
        "type": "commentCreated",
        "commentId": context.commentId,
        "postId": context.postId,
        "textLength": context.textLength
      ]
      
    case .replyCreated(let context):
      return [
        "type": "replyCreated",
        "replyId": context.replyId,
        "commentId": context.commentId,
        "textLength": context.textLength
      ]
      
    case .contentDeleted(let context):
      return [
        "type": "contentDeleted",
        "contentId": context.contentId,
        "contentKind": serializeContentKind(context.kind)
      ]
      
    case .reactionModified(let context):
      var data: [String: Any] = [
        "type": "reactionModified",
        "contentId": context.contentId,
        "contentKind": serializeContentKind(context.contentKind)
      ]
      if let prev = context.previousReaction {
        data["previousReaction"] = serializeReactionKind(prev)
      }
      if let next = context.newReaction {
        data["newReaction"] = serializeReactionKind(next)
      }
      return data
      
    case .pollVoted(let context):
      return [
        "type": "pollVoted",
        "contentId": context.contentId,
        "optionId": context.optionId
      ]
      
    case .contentReported(let context):
      return [
        "type": "contentReported",
        "contentId": context.contentId,
        "reasons": context.reasons.map { serializeReportReason($0) }
      ]
      
    case .gamificationPointsGained(let context):
      return [
        "type": "gamificationPointsGained",
        "points": context.pointsGained,
        "action": serializeGamificationPointsGainedAction(context.action)
      ]
      
    case .gamificationPointsRemoved(let context):
      return [
        "type": "gamificationPointsRemoved",
        "points": context.pointsRemoved,
        "action": serializeGamificationPointsRemovedAction(context.action)
      ]
      
    case .screenDisplayed(let context):
      return [
        "type": "screenDisplayed",
        "screen": serializeScreen(context.screen)
      ]
      
    case .notificationClicked(let context):
      var data: [String: Any] = [
        "type": "notificationClicked",
        "notificationId": context.notificationId
      ]
      if let contentId = context.contentId {
        data["contentId"] = contentId
      }
      return data
      
    case .postClicked(let context):
      return [
        "type": "postClicked",
        "postId": context.postId,
        "source": serializePostClickedSource(context.source)
      ]
      
    case .translationButtonClicked(let context):
      return [
        "type": "translationButtonClicked",
        "contentId": context.contentId,
        "viewTranslated": context.viewTranslated,
        "contentKind": serializeContentKind(context.contentKind)
      ]
      
    case .commentButtonClicked(let context):
      return [
        "type": "commentButtonClicked",
        "postId": context.postId
      ]
      
    case .replyButtonClicked(let context):
      return [
        "type": "replyButtonClicked",
        "commentId": context.commentId
      ]
      
    case .seeRepliesButtonClicked(let context):
      return [
        "type": "seeRepliesButtonClicked",
        "commentId": context.commentId
      ]
      
    case .profileModified(let context):
      var data: [String: Any] = [
        "type": "profileModified",
        "nicknameUpdated": context.nickname.isUpdated,
        "bioUpdated": context.bio.isUpdated,
        "pictureUpdated": context.picture.isUpdated
      ]
      if case .updated(let bioContext) = context.bio {
        data["bioLength"] = bioContext.bioLength
      }
      if case .updated(let pictureContext) = context.picture {
        data["hasPicture"] = pictureContext.hasPicture
      }
      return data
      
    case .groupFollowingChanged(let context):
      return [
        "type": "groupFollowingChanged",
        "groupId": context.groupId,
        "followed": context.followed
      ]

    case .sessionStarted(let context):
      return [
        "type": "sessionStarted",
        "sessionId": context.sessionId
      ]
      
    case .sessionStopped(let context):
      return [
        "type": "sessionStopped",
        "sessionId": context.sessionId
      ]
      
    @unknown default:
      return nil
    }
  }
  
  private static func serializeContentKind(_ kind: OctopusEvent.ContentKind) -> String {
    switch kind {
    case .post: return "post"
    case .comment: return "comment"
    case .reply: return "reply"
    }
  }
  
  private static func serializeReactionKind(_ kind: OctopusEvent.ReactionKind) -> String {
    switch kind {
    case .heart: return "heart"
    case .joy: return "joy"
    case .mouthOpen: return "mouthOpen"
    case .clap: return "clap"
    case .cry: return "cry"
    case .rage: return "rage"
    case .unknown: return "unknown"
    }
  }
  
  private static func serializeReportReason(_ reason: OctopusEvent.ReportReason) -> String {
    switch reason {
    case .hateSpeechOrDiscriminationOrHarassment: return "hateSpeech"
    case .explicitOrInappropriateContent: return "explicit"
    case .violenceAndTerrorism: return "violence"
    case .spamAndScams: return "spam"
    case .suicideAndSelfHarm: return "suicide"
    case .fakeProfilesAndImpersonation: return "fakeProfile"
    case .childExploitationOrAbuse: return "childExploitation"
    case .intellectualPropertyViolation: return "intellectualProperty"
    case .other: return "other"
    }
  }
  
  private static func serializeGamificationPointsGainedAction(_ action: OctopusEvent.GamificationPointsGainedAction) -> String {
    switch action {
    case .post: return "post"
    case .comment: return "comment"
    case .reply: return "reply"
    case .reaction: return "reaction"
    case .vote: return "vote"
    case .postCommented: return "postCommented"
    case .profileCompleted: return "profileCompleted"
    case .dailySession: return "dailySession"
    }
  }
  
  private static func serializeGamificationPointsRemovedAction(_ action: OctopusEvent.GamificationPointsRemovedAction) -> String {
    switch action {
    case .postDeleted: return "postDeleted"
    case .commentDeleted: return "commentDeleted"
    case .replyDeleted: return "replyDeleted"
    case .reactionDeleted: return "reactionDeleted"
    }
  }
  
  private static func serializePostClickedSource(_ source: OctopusEvent.PostClickedSource) -> String {
    switch source {
    case .feed: return "feed"
    case .profile: return "profile"
    }
  }
  
  private static func serializeScreen(_ screen: OctopusEvent.Screen) -> [String: Any] {
    switch screen {
    case .postsFeed(let context):
      var data: [String: Any] = ["type": "postsFeed", "feedId": context.feedId]
      if let relatedTopicId = context.relatedTopicId {
        data["relatedTopicId"] = relatedTopicId
      }
      return data
    case .postDetail(let context):
      return ["type": "postDetail", "postId": context.postId]
    case .commentDetail(let context):
      return ["type": "commentDetail", "commentId": context.commentId]
    case .createPost:
      return ["type": "createPost"]
    case .profile:
      return ["type": "profile"]
    case .otherUserProfile(let context):
      return ["type": "otherUserProfile", "profileId": context.profileId]
    case .editProfile:
      return ["type": "editProfile"]
    case .reportContent:
      return ["type": "reportContent"]
    case .reportProfile:
      return ["type": "reportProfile"]
    case .validateNickname:
      return ["type": "validateNickname"]
    // Unreachable since native iOS 1.13: the settings-list screen was deleted and no
    // emission site for this case survives, though the case itself remains in the
    // native enum. Android still emits it. Kept so the mapping stays total, and so a
    // future iOS screen reusing the case is bridged rather than silently dropped.
    case .settingsList:
      return ["type": "settingsList"]
    case .settingsAccount:
      return ["type": "settingsAccount"]
    case .reportExplanation:
      return ["type": "reportExplanation"]
    case .deleteAccount:
      return ["type": "deleteAccount"]
    // New screens added in native SDK 1.11 — typed support deferred to a follow-up.
    // Surface them as "unknown" so the bridge doesn't crash and JS can detect them.
    case .mainFeed:
      return ["type": "unknown"]
    case .groups:
      return ["type": "unknown"]
    case .groupDetail:
      return ["type": "unknown"]
    // Unified Profile screen added in native SDK 1.13 — the wrapper exposes no
    // Unified Profile surface yet, so it follows the same deferral as above.
    case .otherUserPosts:
      return ["type": "unknown"]
    @unknown default:
      return ["type": "unknown"]
    }
  }
}

File: android/src/main/java/com/octopuscommunity/octopusreactnativesdk/OctopusEventSerializer.kt
package com.octopuscommunity.octopusreactnativesdk

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.octopuscommunity.sdk.domain.model.Gamification
import com.octopuscommunity.sdk.domain.model.Moderation
import com.octopuscommunity.sdk.domain.model.OctopusEvent
import com.octopuscommunity.sdk.domain.model.OctopusItem

/**
 * Serializes OctopusEvent objects to WritableMap for React Native bridge.
 */
object OctopusEventSerializer {

  fun serializeEvent(event: OctopusEvent): WritableMap? {
    val map = Arguments.createMap()
    
    return when (event) {
      is OctopusEvent.PostCreated -> {
        val contentList = event.content.map { content ->
          when (content) {
            OctopusEvent.PostCreated.Content.TEXT -> "text"
            OctopusEvent.PostCreated.Content.IMAGE -> "image"
            OctopusEvent.PostCreated.Content.POLL -> "poll"
          }
        }
        map.putString("type", "postCreated")
        map.putString("postId", event.postId)
        val contentArray = Arguments.createArray()
        contentList.forEach { contentArray.pushString(it) }
        map.putArray("content", contentArray)
        map.putString("topicId", event.groupId)
        map.putInt("textLength", event.textLength)
        map
      }

      is OctopusEvent.CommentCreated -> {
        map.putString("type", "commentCreated")
        map.putString("commentId", event.commentId)
        map.putString("postId", event.postId)
        map.putInt("textLength", event.textLength)
        map
      }

      is OctopusEvent.ReplyCreated -> {
        map.putString("type", "replyCreated")
        map.putString("replyId", event.replyId)
        map.putString("commentId", event.commentId)
        map.putInt("textLength", event.textLength)
        map
      }

      is OctopusEvent.PostDeleted -> {
        map.putString("type", "contentDeleted")
        map.putString("contentId", event.contentId)
        map.putString("contentKind", "post")
        map
      }

      is OctopusEvent.CommentDeleted -> {
        map.putString("type", "contentDeleted")
        map.putString("contentId", event.contentId)
        map.putString("contentKind", "comment")
        map
      }

      is OctopusEvent.ReplyDeleted -> {
        map.putString("type", "contentDeleted")
        map.putString("contentId", event.contentId)
        map.putString("contentKind", "reply")
        map
      }

      is OctopusEvent.ReactionModified -> {
        map.putString("type", "reactionModified")
        map.putString("contentId", event.contentId)
        map.putString("contentKind", serializeContentKind(event.contentKind))
        event.previousReaction?.let {
          map.putString("previousReaction", serializeReactionKind(it))
        }
        event.newReaction?.let {
          map.putString("newReaction", serializeReactionKind(it))
        }
        map
      }

      is OctopusEvent.PollVote -> {
        map.putString("type", "pollVoted")
        map.putString("contentId", event.contentId)
        map.putString("optionId", event.optionId)
        map
      }

      is OctopusEvent.ContentReported -> {
        map.putString("type", "contentReported")
        map.putString("contentId", event.contentId)
        val reasonsArray = Arguments.createArray()
        event.reasons.forEach { reason ->
          reasonsArray.pushString(serializeReportReason(reason))
        }
        map.putArray("reasons", reasonsArray)
        map
      }

      is OctopusEvent.ProfileReported -> {
        map.putString("type", "profileReported")
        map.putString("profileId", event.profileId)
        val reasonsArray = Arguments.createArray()
        event.reasons.forEach { reason ->
          reasonsArray.pushString(serializeReportReason(reason))
        }
        map.putArray("reasons", reasonsArray)
        map
      }

      is OctopusEvent.GamificationPointsGained -> {
        map.putString("type", "gamificationPointsGained")
        map.putInt("points", event.points)
        map.putString("action", serializeGamificationAction(event.action))
        map
      }

      is OctopusEvent.GamificationPointsRemoved -> {
        map.putString("type", "gamificationPointsRemoved")
        map.putInt("points", event.points)
        map.putString("action", serializeGamificationPointsRemovedAction(event.action))
        map
      }

      is OctopusEvent.ScreenDisplayed -> {
        map.putString("type", "screenDisplayed")
        val screenMap = serializeScreen(event)
        map.putMap("screen", screenMap)
        map
      }

      is OctopusEvent.NotificationClicked -> {
        map.putString("type", "notificationClicked")
        map.putString("notificationId", event.notificationId)
        event.contentId?.let {
          map.putString("contentId", it)
        }
        map
      }

      is OctopusEvent.PostClicked -> {
        map.putString("type", "postClicked")
        map.putString("postId", event.postId)
        map.putString("source", when (event.source) {
          OctopusEvent.PostClicked.Source.FEED -> "feed"
          OctopusEvent.PostClicked.Source.PROFILE -> "profile"
        })
        map
      }

      is OctopusEvent.TranslationButtonClicked -> {
        map.putString("type", "translationButtonClicked")
        map.putString("contentId", event.contentId)
        map.putBoolean("viewTranslated", event.viewTranslated)
        map.putString("contentKind", serializeContentKind(event.contentKind))
        map
      }

      is OctopusEvent.CommentButtonClicked -> {
        map.putString("type", "commentButtonClicked")
        map.putString("postId", event.postId)
        map
      }

      is OctopusEvent.ReplyButtonClicked -> {
        map.putString("type", "replyButtonClicked")
        map.putString("commentId", event.commentId)
        map
      }

      is OctopusEvent.SeeRepliesButtonClicked -> {
        map.putString("type", "seeRepliesButtonClicked")
        map.putString("commentId", event.commentId)
        map
      }

      is OctopusEvent.ProfileModified -> {
        val prev = event.previousProfile
        val new = event.newProfile
        val nicknameUpdated = prev?.nickname != new.nickname
        val bioUpdated = prev?.bio != new.bio
        val pictureUpdated = prev?.picture != new.picture
        
        map.putString("type", "profileModified")
        map.putBoolean("nicknameUpdated", nicknameUpdated)
        map.putBoolean("bioUpdated", bioUpdated)
        if (bioUpdated) {
          new.bio?.length?.let { map.putInt("bioLength", it) }
        }
        map.putBoolean("pictureUpdated", pictureUpdated)
        if (pictureUpdated) {
          map.putBoolean("hasPicture", new.picture != null)
        }
        map
      }

      is OctopusEvent.GroupFollowingChanged -> {
        map.putString("type", "groupFollowingChanged")
        map.putString("groupId", event.groupId)
        map.putBoolean("followed", event.followed)
        map
      }

      is OctopusEvent.SessionStarted -> {
        map.putString("type", "sessionStarted")
        map.putString("sessionId", event.sessionId)
        map
      }

      is OctopusEvent.SessionStopped -> {
        map.putString("type", "sessionStopped")
        map.putString("sessionId", event.sessionId)
        map
      }

      else -> null
    }
  }

  private fun serializeContentKind(kind: OctopusItem.ContentKind): String {
    return when (kind) {
      OctopusItem.ContentKind.POST -> "post"
      OctopusItem.ContentKind.COMMENT -> "comment"
      OctopusItem.ContentKind.REPLY -> "reply"
    }
  }

  private fun serializeReactionKind(kind: OctopusItem.Reaction.Kind): String {
    return when (kind) {
      is OctopusItem.Reaction.Kind.Heart -> "heart"
      is OctopusItem.Reaction.Kind.Joy -> "joy"
      is OctopusItem.Reaction.Kind.MouthOpen -> "mouthOpen"
      is OctopusItem.Reaction.Kind.Clap -> "clap"
      is OctopusItem.Reaction.Kind.Cry -> "cry"
      is OctopusItem.Reaction.Kind.Rage -> "rage"
      is OctopusItem.Reaction.Kind.Unknown -> "unknown"
    }
  }

  private fun serializeReportReason(reason: Moderation.ReportReason): String {
    return when (reason) {
      is Moderation.ReportReason.HateSpeechOrDiscriminatoryContent -> "hateSpeech"
      is Moderation.ReportReason.ExplicitOrInappropriateContent -> "explicit"
      is Moderation.ReportReason.ViolenceAndTerrorism -> "violence"
      is Moderation.ReportReason.SpamAndScams -> "spam"
      is Moderation.ReportReason.SuicideAndSelfHarm -> "suicide"
      is Moderation.ReportReason.FakeProfilesAndImpersonation -> "fakeProfile"
      is Moderation.ReportReason.ChildExploitationOrAbuse -> "childExploitation"
      is Moderation.ReportReason.IntellectualPropertyViolation -> "intellectualProperty"
      is Moderation.ReportReason.Other -> "other"
    }
  }

  private fun serializeGamificationAction(action: Gamification.Action): String {
    return when (action) {
      Gamification.Action.POST -> "post"
      Gamification.Action.COMMENT -> "comment"
      Gamification.Action.REPLY -> "reply"
      Gamification.Action.REACTION -> "reaction"
      Gamification.Action.VOTE -> "vote"
      Gamification.Action.POST_COMMENTED -> "postCommented"
      Gamification.Action.PROFILE_COMPLETED -> "profileCompleted"
      Gamification.Action.DAILY_SESSION -> "dailySession"
    }
  }

  private fun serializeGamificationPointsRemovedAction(action: Gamification.Action): String {
    // In Android SDK, GamificationPointsRemoved uses the same enum values as GamificationPointsGained,
    // but the context (removed vs gained) determines the meaning.
    // Map the action values to their "removed" equivalents
    return when (action) {
      Gamification.Action.POST -> "postDeleted"
      Gamification.Action.COMMENT -> "commentDeleted"
      Gamification.Action.REPLY -> "replyDeleted"
      Gamification.Action.REACTION -> "reactionDeleted"
      // For other actions, use the regular serialization (shouldn't happen for removed events)
      else -> serializeGamificationAction(action)
    }
  }

  private fun serializeScreen(event: OctopusEvent.ScreenDisplayed): WritableMap {
    val screenMap = Arguments.createMap()
    // Default set before the `when` below, which stays exhaustive on purpose (a compile-time
    // forcing function: adding a ScreenDisplayed subtype without adding a branch here is a
    // build failure, not a silent gap). This default is the defense-in-depth line for the
    // case that check cannot cover — a native dependency bump paired with a wrapper build that
    // does not also pick up the matching Kotlin subtype — so the bridge still emits a map with
    // a "type" key, `'unknown'`, instead of one missing it entirely.
    screenMap.putString("type", "unknown")

    when (event) {
      is OctopusEvent.ScreenDisplayed.PostsFeed -> {
        screenMap.putString("type", "postsFeed")
        screenMap.putString("feedId", event.feedId)
        event.relatedGroupId?.let {
          screenMap.putString("relatedTopicId", it)
        }
      }
      is OctopusEvent.ScreenDisplayed.PostDetail -> {
        screenMap.putString("type", "postDetail")
        screenMap.putString("postId", event.postId)
      }
      is OctopusEvent.ScreenDisplayed.CommentDetail -> {
        screenMap.putString("type", "commentDetail")
        screenMap.putString("commentId", event.commentId)
      }
      is OctopusEvent.ScreenDisplayed.CreatePost -> {
        screenMap.putString("type", "createPost")
      }
      is OctopusEvent.ScreenDisplayed.Profile -> {
        screenMap.putString("type", "profile")
      }
      is OctopusEvent.ScreenDisplayed.OtherUserProfile -> {
        screenMap.putString("type", "otherUserProfile")
        screenMap.putString("profileId", event.profileId)
      }
      is OctopusEvent.ScreenDisplayed.EditProfile -> {
        screenMap.putString("type", "editProfile")
      }
      is OctopusEvent.ScreenDisplayed.ReportContent -> {
        screenMap.putString("type", "reportContent")
      }
      is OctopusEvent.ScreenDisplayed.ReportProfile -> {
        screenMap.putString("type", "reportProfile")
      }
      is OctopusEvent.ScreenDisplayed.ValidateNickname -> {
        screenMap.putString("type", "validateNickname")
      }
      is OctopusEvent.ScreenDisplayed.SettingsList -> {
        screenMap.putString("type", "settingsList")
      }
      is OctopusEvent.ScreenDisplayed.SettingsAccount -> {
        screenMap.putString("type", "settingsAccount")
      }
      is OctopusEvent.ScreenDisplayed.ReportExplanation -> {
        screenMap.putString("type", "reportExplanation")
      }
      is OctopusEvent.ScreenDisplayed.DeleteAccount -> {
        screenMap.putString("type", "deleteAccount")
      }
      is OctopusEvent.ScreenDisplayed.MainFeed -> {
        screenMap.putString("type", "mainFeed")
        screenMap.putString("feedId", event.feedId)
      }
      OctopusEvent.ScreenDisplayed.Groups -> {
        screenMap.putString("type", "groups")
      }
      is OctopusEvent.ScreenDisplayed.GroupDetail -> {
        screenMap.putString("type", "groupDetail")
        screenMap.putString("groupId", event.groupId)
        // `event.source` (BRIDGE / COMMUNITY) is deliberately not bridged: iOS names the
        // same distinction `clientApp` / `community`, so carrying it needs a wire naming
        // decision of its own rather than an Android-side one.
      }
      // Unified Profile screens added in native SDK 1.13, reachable since the wrapper
      // passes `onNavigateToProfile` (behind `interceptProfileTaps`).
      //
      // `activity` has no iOS counterpart: the native iOS SDK models no separate screen
      // for the connected user's own activity and reports `.profile` for it. The two
      // sides are therefore resolved by *naming* the asymmetry rather than by flattening
      // it — Android keeps the finer tag the native SDK gives it, and `ScreenType`'s
      // TSDoc plus `KNOWN_SCREEN_PLATFORM_GAPS` in `src/__tests__/goldenRoundtrip.test.ts`
      // record that a host counting this action must accept "activity" or "profile".
      // Flattening to "profile" here would instead discard information Android has and
      // silently change tag if iOS ever gains the case.
      is OctopusEvent.ScreenDisplayed.Activity -> {
        screenMap.putString("type", "activity")
      }
      is OctopusEvent.ScreenDisplayed.OtherUserPosts -> {
        screenMap.putString("type", "otherUserPosts")
        screenMap.putString("profileId", event.profileId)
      }
    }
    
    return screenMap
  }
}
